README.txt
 ********************************************************
 *                                                      *
 * Efficient Subwindow Search (ESS) implemention in C++ *
 * entry point and commmand line interface              *
 *                                                      *
 * version 1.1, August 4th 2008                         *
 *   included interface to call from SVMstruct          *
 *                                                      *
 *   Copyright 2006-2008 Christoph Lampert              *
 *   Contact: <mail@christoph-lampert.org>              *
 *                                                      *
 *  Performs a branch-and-bound search for an arbitrary *
 *  quality function, provided a routine to bound it:   *
 *  see  C. H. Lampert, M. B. Blaschko and T. Hofmann:  *
 *       Beyond Sliding Windows: Object Localization    *
 *       by Efficient Subwindow Search, CVPR (2008)     *
 *                                                      *
 *  For structured regression training, see             *
 *       M. B. Blaschko and C. H. Lampert               *
 *       Learning to Localization Objects With          *
 *       Structured Output Regression, ECCV (2008)      *
 *                                                      *
 ********************************************************

This is sample implementation of Efficient Subwindow Search,
a method to replace sliding windows approaches to object
localization by a branch-and-bound search. 

The default quality function is to use a spatial pyramid kernel 
with levels of size 1x1, 2x2 ... NxN where N can be chosen at 
runtime. Default is N=1, i.e. bag-of-visual-words model. 


USAGE: 

./ess imagewidth imageheight weight-file data-file

The weight-file is an ASCII file of 1 entry per line. 
It contains the concatenation of the per-cluster weights for 
each pyramid level. The levels are ordered by increasing index.
Within each level, weights are aranged in reading order. All 
weights must be present, as the number of cluster centers 
is calculated by dividing the total number of weights by the
total number of pyramid cells, which is N*(N+1)*(2N+1)/6. 

The data-file is an ASCII File of 3 columns. Each line has the format

X Y CLST

where (X,Y) are the coordinates of a feature in the image and CLST is
an ID attached to it, e.g. it's ID in a codebook representation.
The ID and X,Y are used to look up the correct weight for this feature 
point in a candidate box.

Output is a single line with all detected box in the format

SCORE LEFT TOP RIGHT BOTTOM 

repeating as often as there are boxes.


The main routine 

Box pyramid_search(int argnumpoints, int argwidth, int argheight, 
                   double* argxpos, double* argypos, double* argclst,
                   int argnumclusters, int argnumlevels, double* argweight)

can also be called from an external application. Extracted multiple boxes 
requires a loop, see the main() routine.



EXAMPLE FILES: 

examples/cow.weight is a weightfile with 3000 clusters in 1 level.
examples/car-l1.weight is a weightfile with 1000 clusters in 1 level.
examples/car-l2.weight is a weightfile with 1000 clusters in 2 levels, 
making it 5000 weights in total.

examples/cow.clst is a data-file of 13498 feature points.
examples/cow.clst is a data-file of 13498 feature points.

Arguments can be passed to the executable by environment variables:

numlevels=2 ./ess 151 101 examples/car-l2.weight examples/car.clst

returns the best box for a 2-level pyramid. 

maxresults=3 ./ess 151 101 examples/car-l1.weight examples/car.clst

returns the 3 best car detections instead of just 1. Both arguments
can be combined:

maxresults=2 numlevels=2 ./ess 151 101 examples/car-l2.weight examples/car.clst


Outputs for the examples are:

./ess 368 272 examples/cow.weight examples/cow.clst
# 10978.8779297 89 117 258 209 
	
./ess 151 101 examples/car-l1.weight examples/car.clst
# 1.51183950901 39 56 118 77 0.286267101765 4 61 144 65 0.138369500637 53 88 147 90 

numlevels=2 ./ess 151 101 examples/car-l2.weight examples/car.clst
# 1.69141745567 33 55 115 77 

maxresults=2 numlevels=2 ./ess 151 101 examples/car-l2.weight examples/car.clst
# 1.69141745567 33 55 115 77 0.15714520216 4 60 136 65 



TEST FILES: 

Apart from the examples, there is one synthetic test case so far: 

maxresults=4 ./ess 5 5 examples/test_corners.weight examples/test_corners.clst

should return 

# 1.39999997616 4 4 4 4 1.29999995232 4 0 4 1 1.20000004768 1 4 1 4 1.10000002384 1 1 1 1 



Changelog.txt
03/08/2008 v1.2: added code for structure regression training and subwindow module

07/04/2008 v1.1: added code to compile as library and added Python Wrapper

24/03/2008 v1.0: Initial Release 


ess.cc
/********************************************************
 *                                                      *
 * Efficient Subwindow Search (ESS) implemention in C++ *
 * entry point and commmand line interface              *
 *                                                      *
 *  Performs a branch-and-bound search for an arbitrary *
 *  quality function, provided a routine to bound it,   *
 *  see  C. H. Lampert, M. B. Blaschko and T. Hofmann:  *
 *       Beyond Sliding Windows: Object Localization    *
 *       by Efficient Subwindow Search, CVPR (2008)     *
 *                                                      *
 *   Copyright 2006-2008 Christoph Lampert              *
 *   Contact: <mail@christoph-lampert.org>              *
 *                                                      *
 *  Licensed under the Apache License, Version 2.0 (the *
 *  "License"); you may not use this file except in     *
 *  compliance with the License. You may obtain a copy  *
 *  of the License at                                   *
 *                                                      *
 *     http://www.apache.org/licenses/LICENSE-2.0       *
 *                                                      *
 *  Unless required by applicable law or agreed to in   *
 *  writing, software distributed under the License is  * 
 *  distributed on an "AS IS" BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, either express or        *
 *  implied. See the License for the specific language  *
 *  governing permissions and limitations under the     *
 *  License.                                            *
 *                                                      *
 ********************************************************/

#include <cassert>
#include <iostream>
#include <iomanip>
#include <fstream>
#include <vector>
#include <string>

#include "ess.hh"
#include "quality_pyramid.hh"

#define MAXDATAPOINTS 100000 // ad hoc limits
#define MAXWIDTH 8192
#define MAXHEIGHT 8192
#define MAXCLUSTERS 100000

static int maxresults = 1;
static int numlevels = 1;
static int maxiterations = 10000000;
static int verbose = 0;

// Here we chose the class to calculate quality bounds for us.
// It has to have at least the interface of the QualityFunction class.
//
// We use 'PyramidQualityFunction', because it's flexible.
// We use 'BoxQualityFunction' is a little easier to set up.
static PyramidQualityFunction quality_bound;

// For 'most_violated_constraint' functionality in structured search,
// we need to know the ground truth boxes to calculate the loss 
static double* ref_boxes;
static int num_ref_boxes;

// central routine during branch-and-bound search:
// 1) extract the most promising candidate region 
// 2) split it, if necessary 
// 3) calculate upper bounds for the parts
// 4) re-insert the parts

static int extract_split_and_insert(sstate_heap *pH) {

    // step 1) find the most promising candidate region 
    const sstate* curstate = pH->top();

    // step 2a) check if the stop criterion is reached
    const int splitindex = curstate->maxindex();
    if (splitindex < 0)
        return -1;    // no more splits => convergence

    // step 2b) otherwise, create two new states as copies of the old, except for splitindex
    pH->pop();
    sstate* newstate0 = new sstate(*curstate);
    newstate0->high[splitindex] = (curstate->low[splitindex] + curstate->high[splitindex])>>1;

    sstate* newstate1 = new sstate(*curstate);
    newstate1->low[splitindex] = (curstate->low[splitindex] + curstate->high[splitindex]+1)>>1;

    // the old state isn't needed anymore
    delete curstate; curstate=NULL;
    
    // step 3&4) calculate upper bounds for the parts and reinject them 
    if ( newstate0->islegal() ) {
        newstate0->upper = quality_bound.upper_bound(newstate0);
        pH->push(newstate0);
    }
    if ( newstate1->islegal() ) {
        newstate1->upper = quality_bound.upper_bound(newstate1);
        pH->push(newstate1);
    }
    
    // no error, but also no convergence, yet
    return 0;
}

extern "C" {

// main entry site for efficient subwindow search.
// performs preprocessing and then branch-and-bound
// We make it "extern C", so it's easier to call e.g. from Python
//
// INPUT: int argnumpoints,     : number of data points
//        int width, height     : width and height of image
//        double* argxpos,      : x-coordinate of every point
//        double* argypos,      : y-coordinate of every point
//        double* argclst       : cluster ID of each point (starts at 0)
//        int argnumclusters,   : number of clusterIDs 
//        int argnumlevels,     : number of levels in the pyramid 
//        double* weightsdata   : vector of cluster weights
// OUTPUT: Box outputBox        : box in [left,top,right,bottom,score] format

Box pyramid_search(int argnumpoints, int argwidth, int argheight, 
                   double* argxpos, double* argypos, double* argclst,
                   int argnumclusters, int argnumlevels, double* argweight) {
    Box outputBox = {-1, -1, -1, -1, -1.};
    argwidth += 1; // make space for 1 pixel padding
    argheight += 1;

// set up structure for pyramid grid parameters
// TODO: find a nicer way to handle the variable number of parameters
    const int numcells = argnumlevels*(argnumlevels+1)*(2*argnumlevels+1)/6;

    PyramidParameters paramstruct;
    paramstruct.numlevels=argnumlevels;
    paramstruct.weightptr = new double*[numcells];
    for (unsigned int i=0; i < numcells; i++) {
        paramstruct.weightptr[i] = &argweight[i*argnumclusters];
    }
    
// set up everything needed to calculate qualities and bounds
    quality_bound.setup(argnumpoints, argwidth, argheight, argxpos, argypos, argclst, &paramstruct);
    delete [] paramstruct.weightptr;

// intialize the search space (start with full image)
    sstate* fullspace = new sstate(argwidth, argheight);
    
// push first box set into priority queue
    sstate_heap H;
    H.push(fullspace);

// main loop. Iterate extract/split/evaluate/reinsert until convergence or forced exit
    long counter=1;
    while ((extract_split_and_insert(&H) >= 0) && (counter < maxiterations)) {
        if (verbose) {
            if ((counter % verbose) == 0) {
                const sstate *curmax = H.top();
                std::cerr << "#counter " << std::setw(8) << counter;
                std::cerr << " heapsize " << std::setw(8) << H.size();
                std::cerr << " <" << std::setw(4) << curmax->upper << " > ";
                std::cerr << curmax->tostring();
            }
        }
        counter++;
    }
// at convergence or error, return result or best guess
    const sstate* curstate = H.top();
    outputBox.left   = ((curstate->low[0]+curstate->high[0])>>1) -1;  // remove padding
    outputBox.top    = ((curstate->low[1]+curstate->high[1])>>1) -1;
    outputBox.right  = ((curstate->low[2]+curstate->high[2])>>1) -1;
    outputBox.bottom = ((curstate->low[3]+curstate->high[3])>>1) -1;
    outputBox.score  = curstate->upper;

// Free memory of the states that are still in the queue
        while (!H.empty()) {
        delete H.top();
            H.pop();
    }

// generic function to free any internal resource 
    quality_bound.cleanup();

    return outputBox;
}



// entry site for 'find_most_violated_constraint' in structured search
// it adds the loss-function in scoring the boxes, apart from that it's
// the same as pyramid_search(...)
//
// INPUT: int argnumpoints,     : number of data points
//        int width, height     : width and height of image
//        double* argxpos,      : x-coordinate of every point
//        double* argypos,      : y-coordinate of every point
//        double* argclst       : cluster ID of each point (starts at 0)
//        int argnumclusters,   : number of clusterIDs 
//        int argnumlevels,     : number of levels in the pyramid 
//        double* weightsdata   : vector of cluster weights
//        int num_gt_boxes      : number of ground truth boxes to use in loss
//        double* weightsdata   : coordinates of ground truth boxes in 1D vector
//
// OUTPUT: Box outputBox        : box in [left,top,right,bottom,score] format
Box pyramid_search_structured(int argnumdatapts, int argwidth, int argheight, 
                   double* argxpos, double* argypos, double* argclst,
                   int argnumclusters, int argnumlevels, double* argweightdata, 
                   int num_gt_boxes, double* gt_box) {

        ref_boxes = gt_box;
        num_ref_boxes = num_gt_boxes;
        Box outputBox = pyramid_search(argnumdatapts, argwidth, argheight, 
                                       argxpos, argypos, argclst, 
                                       argnumclusters, argnumlevels, argweightdata);
        return outputBox;
}

} // end of extern "C"


#ifdef __MAIN__

static void usage(char *progname) {
    std::cerr << "usage: " << progname << " width height weight-file data-file\n";
    exit(1);
}

// helper routine to read ASCII data files 
// INPUT: filename: name of input file
// INPUT/OUTPUT: data: collection of vectors to be filled
static int readdata_NxM(const char* filename, std::vector< std::vector<double> > &data) {

    const unsigned int numcolumns=data.size();

// start with empty vectors
    for (unsigned int i=0;i<numcolumns;i++) {
        data[i].clear();
    }

    std::ifstream infile(filename);
    if (!infile)
        return -1;

    while (! infile.eof() ) {
        for (unsigned int i=0;i<numcolumns;i++) {
            double tmpval;
            infile >> tmpval;
            data[i].push_back(tmpval);
        }
    }
    infile.close();

// usually, eof occurs too late: we have read one extra entry
    for (unsigned int i=0;i<numcolumns;i++)
        data[i].pop_back();

// all vector should have same length
        const int numpts = data[0].size();
    return numpts;
}

// parse an int value from env variable
static int igetenv(const char* name, int defaultvalue, int low, int high) {
    int val = defaultvalue;
    if (getenv(name))
        return atoi(getenv(name));
    if (val < low)
        val=low;
    else if (val>high)
        val=high;
    return val;
}

// convenience function to control the behaviour through environment variables
static void set_parameters() {
    maxresults = igetenv("maxresults",1,1,10000);
    numlevels = igetenv("numlevels",1,1,100);
    maxiterations = igetenv("iterations",1,100000000,100000000);
    verbose = igetenv("verbose",0,0,100000000);
    return;
}

int main(int argc, char* argv[]) {
    if (argc < 5)
        usage(argv[0]);

    set_parameters();
    sstate_heap H;

// first two arguments are width and height. 
    const int width = atoi(argv[1]);
    const int height = atoi(argv[2]);
    if ((width<2) || (width>MAXWIDTH) || (height<2) || (height>MAXHEIGHT))
       usage(argv[0]);

// read weight for clusters (1 column)
    std::vector< std::vector<double> > weightdata(1);
    int numweights = readdata_NxM(argv[3], weightdata);
    if (numweights <= 0) {
        std::cerr << "Error reading data from file \n" << argv[3] << std::endl;
        usage(argv[0]);
    }
    
    const int numcells = numlevels*(numlevels+1)*(2*numlevels+1)/6; // = 1+2^2+...+n^2
    const int numclusters = numweights/numcells;
    if (numclusters > MAXCLUSTERS) {
        std::cerr << "Can't handle that many clusters" << std::endl;
        usage(argv[0]);
    }


// read 3-column data file in format x,y,clusterID
    std::vector< std::vector<double> > rawdata(3);
    int datapts = readdata_NxM(argv[4], rawdata);
    if (datapts <= 0) {
        std::cerr << "Error reading data from file \n" << argv[4] << std::endl;
        usage(argv[0]);
    }

// convert from Nx3 matrix to individual vectors. Not very elegant...
    double* xpos = &rawdata[0][0];
    double* ypos = &rawdata[1][0];
    double* clst = &rawdata[2][0];
    double* weights = &weightdata[0][0];

// loop over target number of boxes
    for (int k=maxresults; k>0; --k) {
        Box bestBox = pyramid_search(datapts, width, height,  xpos, ypos, clst, 
                                     numclusters, numlevels, weights);
        std::cout << std::setprecision(12) << bestBox.score << " ";
        std::cout << bestBox.left << " ";
        std::cout << bestBox.top << " ";
        std::cout << bestBox.right << " ";
        std::cout << bestBox.bottom << " " ;

// before searching for next boxes, zero out the region of the box found.
// because removing elements from an array is cumbersome, we create a new one 
// from only points outside the box
        if (k>0) {
            std::vector< std::vector<double> > tmpdata(3);
            for (int i=0;i<datapts;i++) {
                if  ((xpos[i]<bestBox.left) || (xpos[i]>bestBox.right)
                    || (ypos[i]<bestBox.top) || (ypos[i]>bestBox.bottom)) {
                        tmpdata[0].push_back(xpos[i]);
                        tmpdata[1].push_back(ypos[i]);
                        tmpdata[2].push_back(clst[i]);
                }
            }
            datapts = tmpdata[0].size();
            for (int i=0;i<3;i++) {
                rawdata[i].resize(datapts);
                std::copy(tmpdata[i].begin(), tmpdata[i].end(), rawdata[i].begin());
            }
        }
    }
    std::cout << std::endl;
}
#endif


subwindow.py
"""A module for SVM^python for object localization."""

# Christoph Lampert, chl@tuebingen.mpg.de, 2007-2008
#
# for descriptions of the learning see
# [1] Blaschko, Lampert: "Object Localization by Structured Regression", ECCV 2008
# for the branch-and-bound method to find the most violated constraint:
# [2] Lampert, Blaschko, Hoffman: "Beyond Sliding Windows: ..." CVPR 2008

import sys
import svmapi

# prepare numeric stuff for branch-and-bound search
from ctypes import c_int, c_double, Structure, byref
import pylab
import numpy
from numpy import array,ones,histogram,hstack
from numpy.ctypeslib import load_library,ndpointer
import numpy
numpy.seterr("raise") 

class Box_struct(Structure):
        """Structure to hold left,top,right,bottom and score of a box instance.
           The fields have to coincide with the C-version in pyramid_search.h
        """
        _fields_ = [("left", c_int), 
                    ("top", c_int), 
                    ("right", c_int), 
                    ("bottom", c_int), 
                    ("score", c_double) ]

try:
    pyramidlib = load_library("libess.so","/kyb/agbs/chl/libs/")
except OSError:
    pyramidlib = load_library("libess.so","/agbs/cluster/chl/libs/")

pyramidlib.pyramid_search_structured.restype = Box_struct
pyramidlib.pyramid_search_structured.argtypes = [c_int,c_int,c_int,     # numpoints, width, heigth,
        ndpointer(dtype=c_double, ndim=1, flags='C_CONTIGUOUS'),        # argxpos
        ndpointer(dtype=c_double, ndim=1, flags='C_CONTIGUOUS'),        # argypos
        ndpointer(dtype=c_double, ndim=1, flags='C_CONTIGUOUS'),        # argclst
        c_int, c_int,                                                   # numclusters, numlevels
        ndpointer(dtype=c_double, ndim=1, flags='C_CONTIGUOUS'),        # 1D weight vector
        c_int, ndpointer(dtype=c_double, ndim=1, flags='C_CONTIGUOUS')] # ground truth size, ground truth box

class training_image:
    def __init__(self, imageid, xyc):
        self.id = imageid
        self.x = array([int(val) for val in xyc[:,0]], c_double)
        self.y = array([int(val) for val in xyc[:,1]], c_double)
        self.c = array([int(val) for val in xyc[:,2]], c_double)
        self.n = xyc.shape[0]
        self.width = int(max(self.x))+1
        self.height = int(max(self.y))+1

def array_to_Box(coords=[0,0,0,0], score=0.):
        return Box_struct(coords[0],coords[1],coords[2],coords[3],score)

def Box_to_array(box):
        box_values = []
        if isinstance(box, list):
            for b in box:
                box_values.extend([b.left,b.top,b.right,b.bottom,b.score])
        else:
            box_values.extend([box.left,box.top,box.right,box.bottom,box.score])
        return array(box_values, c_double)

def parse_parameters(sparm):
    """Sets attributes of sparm based on command line arguments.
    
    This gives the user code a chance to change sparm based on the
    custom command line arguments.  The custom command line arguments
    are stored in sparm.argv as a list of strings.  The custom command
    lines are stored in '--option', then 'value' sequence.
    
    If this function is not implemented, any custom command line
    arguments are ignored and sparm remains unchanged."""
    sparm.numbins = 3000    # number of bag-of-visual-word codebooks
    sparm.numlevels = 1     # no spatial pyramid 


def read_examples(filename, sparm):
    """Parses an input file into an example sequence.
       Currently, we parse only raw-test-file in the VOC format:

       # ID   left top right bottom
       where [left,top,right,bottom] is one box instance, or just
       # ID
       if there is no box.
       
       e.g. 
       005181 224 255 246 276
       005181 90 238 112 258
       005194 151 88 455 418
       005259 72 265 124 298
       005259 124 259 166 298
       005259 165 267 191 294
       000002
       000003
       000005
   
       Note that duplicate IDs are allowed for the positive class.
       Those are combined into single examples with multiple labels,
       should can be handle by the delta-loss.
       
       We read only the IDs and the feature files themselves, since 
       they are only needed in the branch-and-bound C-routine anyway.
   """
    QUANT_FACTOR=1 # quantize all coordinated by this (extra speedup)
    
    # This routine reads box files of the VOC type
    examples = []
    lastid = -1

    # Open the file and read each example.
    path = 'train/'
    for line in file(filename):
        # Get rid of comments.
        if line.find('#'): 
            line = line[:line.find('#')]
        tokens = line.split()
        # If the line is empty, who cares?
        if not tokens: 
            continue

        # Get the ID (goes into features) and box (goes into target)
        imageid = int(tokens[0])
        if (imageid != lastid):
            try: 
                xyc = pylab.load(path + '%06d.clst' % imageid)
                xyc[:,0] /= QUANT_FACTOR
                xyc[:,1] /= QUANT_FACTOR
            except IOError:
                path = 'test/'
                xyc = pylab.load(path + '%06d.clst' % imageid)
                xyc[:,0] /= QUANT_FACTOR
                xyc[:,1] /= QUANT_FACTOR
    
        examples.append( [training_image(imageid, xyc), []] ) # automatic xyc split, no label(s)
        lastid = imageid

        target = [int(x)/QUANT_FACTOR for x in tokens[1:]]
        if (len(target) == 4):
            target=array_to_Box(target,1.)      # we use the score as indicator whether it's a positive box
            examples[-1][1].append(target)
        else:
            width=examples[-1][0].width
            height=examples[-1][0].height
            target=array_to_Box([0,0,0,0],-1.)  # '-1' means it's a negative full image
            examples[-1][1].append(target)
            
    # Print out some useful statistics.
    print len(examples), 'examples read'
    return examples


def init_model(sample, sm, sparm):
    """Initialize the size of bins for the weight vector."""
    # 
    sm.size_psi = sparm.numbins   # for a spatial pyramid, size_psi is total histogram length
    print 'size_psi set to',sm.size_psi


def classification_score(ex,y,sm,sparm):
    """Return score for (example, label) pair. Unclear if this is in active use"""
    # Utilize the svmapi.Model convenience method 'classify'. 
    # 
    return sm.svm_model.classify(psi(ex,y,sm,sparm))


def classify_example(ex, sm, sparm):
    """Returns the classification of an example 'x'. This routine is only used during testing."""

    # We should make this a command-line option, or determine it from somewhere.
    # We have to set this 'again', because this part is used in testing, the other in training.
    sparm.numbins=3000
    sparm.numlevels=1
    best_box = find_most_violated_constraint(ex, [], sm, sparm)

    # For the VOC ranking task, returning a box with bad score is never worse than not 
    # returning it. If the aim to classification, we might want to return 'no box' if the
    # score is 0 or less.
    return best_box


def find_most_violated_constraint(ex,y,sm,sparm):
    """Returns the most violated constraint for example (ex,box)."""
    # This is the crucial routine for the max-margin training with a joint kenel map. 
    # For object localization, we can't do the search exhaustively. Instead, we 
    # call the efficient branch-and-bound search from [1].

    # w should be an array anyway, but making sure doesn't hurt
    w = array([val for val in sm.w], c_double)

    # convert the ground truth box-structures to a numpy array
    ground_truth_boxes = Box_to_array(y)
    num_ground_truth_boxes=0
    if len(ground_truth_boxes)>0:
        if ground_truth_boxes[4]>0:
            num_ground_truth_boxes = len(ground_truth_boxes)/5

    #print ground_truth_boxes
    # do the search itself.
    box = pyramidlib.pyramid_search_structured(ex.n, ex.width, ex.height, ex.x, ex.y, ex.c, \
                     sparm.numbins, sparm.numlevels, w, num_ground_truth_boxes, ground_truth_boxes)

    # Here's a good place to check if both the implementation of 'most_violated_constraint' coincide.
    #
    #assert( abs(classification_score(ex, box, sm, sparm)+loss(y,box,sparm)-box.score) < 0.0001 )
    #score1 = classification_score(ex, box, sm, sparm)
    #loss1 = loss(y,box,sparm)
    #if abs(score1+loss1-box.score) >= 0.0001:
    #  print "Warning! Classifcation score differs from box.score: %f+%f != %f" % (score1,loss1,box.score)
    #print >> sys.stderr, score1,loss1, score1+loss1, " : " ,
    #write_label(sys.stderr,box)
    return box


def psi(ex, y, sm, sparm):
    """Returns the combined feature vector Psi(x,y).
       For us this is the histogram of the image x restricted to the box y.
       y should derive from 'find_most_violated_costraint' and be a real box. 
       If it's a list, we use just the first box inside. If the list is empty,
       we use the whole image (used in testing).
    """

    if isinstance(y, list): # in theory, this should only be called with single boxes as y.
         y=y[0]             # but just in case, for a list of boxes we use only the first one
    
    # Crop the image ex to the region given by y
    idx = (ex.x >= y.left) & (ex.x <= y.right) & (ex.y >= y.top) & (ex.y <= y.bottom)
    clsthist = histogram(ex.c[idx], xrange(sparm.numbins))[0]
    return svmapi.Sparse(clsthist)


def init_constraints(sample, sm, sparm):
    """Initializes special constraints, if necessary. 

    The constraints "F(x, box ) - F(x,no box) > 1-xi" for all x with a 'real' box remain
    all through-out learning. We add them to the constraint set once in the beginning. 
    This also fixes some degenerate behaviour of SVMstruct when starting with w==0.
    """
    import svmapi
    constraints = []
    curslack = len(sample)+1
    for feature,boxlist in sample:
        for box in boxlist:
            if box.score < 0:
               continue # use images which contain an object instance

            feature_hist = psi(feature, box, sm, sparm)
            sparse = svmapi.Sparse( feature_hist )
            lhs = svmapi.Document([sparse], costfactor=1, slackid= curslack+100000)

            # Append the lhs and the rhs to the constraint set
            constraints.append((lhs, 1.))
    return constraints


# Boxes _include_ their boundaries, i.e.g even a box with left==right and top==bottom 
# still has size 1. Only illegal boxes (left>right) or (top>bottom) can have size 0 or less.
def box_area(box):
        return (box.right-box.left+1)*(box.bottom-box.top+1)

# Box overlap is box_area(intersection)/area(union). The 'union' is not a rectangle, but
# luckily we can calculate area(union) = box_area(box1)+box_area(box2)-box_area(intersection)
def box_overlap(box1, box2):
        intersection = array_to_Box([ max(box1.left,box2.left), max(box1.top,box2.top), \
                                      min(box1.right,box2.right), min(box1.bottom,box2.bottom) ])
        if (intersection.left<=intersection.right) & (intersection.top<=intersection.bottom):
                return box_area(intersection)/float(box_area(box1)+box_area(box2)-box_area(intersection))
        else:
                return 0.

def loss_function(overlap):
        return (1.-overlap)     # simplest choice
        #return max(1.-2*overlap,0.) # 50% overlap tolerance region 
        #if overlap>=0.5:    # 0-1 loss as VOC scores with 
        #   return 0
        #else:
        #   return 1

def loss(y, ybar, sparm):
    """Loss function for label pair (y,ybar).
       In training, y is a list of ground truth boxes. ybar is a single candiate box.
       For 'positive' examples, the loss is 1 minus the area overlap of the boxes. 
       For 'negative' examples, the loss is the box overlap with the full image.
       In testing, y is empty, and the loss is zero.
    """

    if (y == []): # a test time there are no ground truth boxes => Delta=0
       return 0.

    if (y[0].score < 0. ): # negative example: Delta(box, nobox)=1 (don't predict a box)
        return 1.

    # positive example: it's enough if the prediction overlaps _one_ of the ground truth boxes
    maxoverlap = max([ box_overlap(box, ybar) for box in y])
    return loss_function(maxoverlap)


def write_label(fileptr, y):
    """Write a predicted label to an open file.
       We use the usual format for VOC prediction task (except for the ID)
    """
    print>>fileptr, y.score, y.left, y.top, y.right, y.bottom


def print_iteration_stats(ceps, cached_constraint, sample, sm,
                          cset, alpha, sparm):
    """Called after each iteration. Currently, we don't do anything, but 
       we could e.g. use it to write the best model learned so far to disk."""

    #import cPickle, bz2
    #filename = 'temp.model' # unfortunately, we don't know a good filename 
    #cPickle.dump(sm, bz2.BZ2File(filename, 'w') )

